### ISSUE TITLE
Add MixPanel event for logging out with data

### ISSUE THREAD

**Issue by githubbob42** [githubbob42/mingle2github2#2977]
**[Mingle Card: 3274](https://liquidframeworks.mingle.thoughtworks.com/projects/alpine_mobile/cards/3274)**
```
| | |
|-|-|\* added logic to include whether the user logged out with unsynced data to pass down to web server to log to mixpanel.  

\* Leveraging Maurya's logout middleware that writes an event to mixpanel that a user logged out.  Just adding a property on to that event.||[Push to Site fxqa1](http://buildslave3.basecamp.liquidframeworks.com:5000/requestQAbuild?qaSite=fxqa1&repoUrl=https%3A%2F%2Fgithub.com%2Fgithubbob42%2Falpine-mobile&branch=card3274_mixpanelLogout&repoName=alpine-mobile&repoOwner=Liquidframeworks&cards=%5B%7B%22projectName%22%3A%22alpine_mobile%22%2C%22cardNumber%22%3A%223274%22%7D%5D&requestNumber=1432&requestUrl=https%3A%2F%2Fgithub.com%2FLiquidframeworks%2Falpine-mobile%2Fpull%2F1432)||[Push to Site fxqa2](http://buildslave3.basecamp.liquidframeworks.com:5000/requestQAbuild?qaSite=fxqa2&repoUrl=https%3A%2F%2Fgithub.com%2Fgithubbob42%2Falpine-mobile&branch=card3274_mixpanelLogout&repoName=alpine-mobile&repoOwner=Liquidframeworks&cards=%5B%7B%22projectName%22%3A%22alpine_mobile%22%2C%22cardNumber%22%3A%223274%22%7D%5D&requestNumber=1432&requestUrl=https%3A%2F%2Fgithub.com%2FLiquidframeworks%2Falpine-mobile%2Fpull%2F1432)||[Push to Site fxqa3](http://buildslave3.basecamp.liquidframeworks.com:5000/requestQAbuild?qaSite=fxqa3&repoUrl=https%3A%2F%2Fgithub.com%2Fgithubbob42%2Falpine-mobile&branch=card3274_mixpanelLogout&repoName=alpine-mobile&repoOwner=Liquidframeworks&cards=%5B%7B%22projectName%22%3A%22alpine_mobile%22%2C%22cardNumber%22%3A%223274%22%7D%5D&requestNumber=1432&requestUrl=https%3A%2F%2Fgithub.com%2FLiquidframeworks%2Falpine-mobile%2Fpull%2F1432)||[Push to Site fxqa4](http://buildslave3.basecamp.liquidframeworks.com:5000/requestQAbuild?qaSite=fxqa4&repoUrl=https%3A%2F%2Fgithub.com%2Fgithubbob42%2Falpine-mobile&branch=card3274_mixpanelLogout&repoName=alpine-mobile&repoOwner=Liquidframeworks&cards=%5B%7B%22projectName%22%3A%22alpine_mobile%22%2C%22cardNumber%22%3A%223274%22%7D%5D&requestNumber=1432&requestUrl=https%3A%2F%2Fgithub.com%2FLiquidframeworks%2Falpine-mobile%2Fpull%2F1432)||[Push to Site fxsignoff1](http://buildslave3.basecamp.liquidframeworks.com:5000/requestQAbuild?qaSite=fxsignoff1&repoUrl=https%3A%2F%2Fgithub.com%2Fgithubbob42%2Falpine-mobile&branch=card3274_mixpanelLogout&repoName=alpine-mobile&repoOwner=Liquidframeworks&cards=%5B%7B%22projectName%22%3A%22alpine_mobile%22%2C%22cardNumber%22%3A%223274%22%7D%5D&requestNumber=1432&requestUrl=https%3A%2F%2Fgithub.com%2FLiquidframeworks%2Falpine-mobile%2Fpull%2F1432)||[Push to Site fxsignoff2](http://buildslave3.basecamp.liquidframeworks.com:5000/requestQAbuild?qaSite=fxsignoff2&repoUrl=https%3A%2F%2Fgithub.com%2Fgithubbob42%2Falpine-mobile&branch=card3274_mixpanelLogout&repoName=alpine-mobile&repoOwner=Liquidframeworks&cards=%5B%7B%22projectName%22%3A%22alpine_mobile%22%2C%22cardNumber%22%3A%223274%22%7D%5D&requestNumber=1432&requestUrl=https%3A%2F%2Fgithub.com%2FLiquidframeworks%2Falpine-mobile%2Fpull%2F1432)| |
```

Narrative
---------

Acceptance Criteria

When the user logs out and they have items in the upload queue then send an event to MixPanel indicating that the user logged out with data.

If the user is logging out and they do not have any data in the upload queue then send an event to MixPanel indicating that the user simply logged out.

I'm assuming that we will automatically get the ORG, user, date and time.

**Send this MixPanel event directly, not through the sync packet. Since they are logging out, we know that they must be online.**  (see first line of Analysis)

This is the do-over card. Finish card #3311 before working on this card! Card #3311 should be merged into master before starting this card.

Keep this card very simple and only make the mixpanel event. With #3311 and pr #1399 you should have everything you need to make this a very simple and small PR. If that is not the case, then lets talk about it before you work on this.

Historical Data:
----------------

Analysis
--------

The one problem I have uncovered is that it looks like we'll need to include our mixpanel ID in the client code.

BONUS: Maurya just mentioned that he already added a logout mixpanel event (PR: https://github.com/Liquidframeworks/alpine-mobile/pull/1300)  but it doesn't cover this particular use case where we want to track if the user logged out with data.  I should be able to piggy back on his PR and simply pass the logged-out-with-data property with the request... CC: Not quite that simple - maurya's PR doesn't make a request to log out - he just added middleware to publish an event on an existing route.

BR: Correct, the logout is handled on the existing route via respondBackOnRevokewhich calls connection.revokeOAuth.  I simply want to pass some additional request parameters along to the logout method (that Maurya added) that adds the mixpanel entry.

Note that every time a user logs in, there is at least one entry in the eventstore table for "sync:complete".  This means that EVERY time we click the Logout button, we ALWAYS see the red button warning that we will delete data.  And that unless we search the events for model events we really don't know if the user is logging out with unsynced data.  DP also just mentioned that attachments are not stored in the eventstore so we'll to check that as well.

Craig just submitted a PR (https://github.com/Liquidframeworks/alpine-mobile/pull/1386) that checks for attachment for the sync changes list.  I don't think that I can leverage Craig's PR directly but I can do essentially the same thing that he is doing, except that I need to check for ALL model events, not just new/created events (and will need the actual events and not human readable representations - per card #3276). CC: Should these views be driven off of the same model, or is the overlap a coincidence? Is there a need for a list of all pending changes that could be consumed by both views (albeit transformed/projected/aggregated if necessary)?

BR:  I thought about this too and wanted to discuss that as an option with you.  I updated my branch with a way to handle this if you want to take a look...  Alfter talking it over we decided to keep them separate (overlap is a coincidence).  If we find a pattern start forming between to two as time goes on we'll consider merging the two into a new viewmodel.

In regards to the "always prompting for delete" because there is always at least a "sync:complete" event in the eventstore, I suggest that we remove the check for that, surrounding the prompt, to simply always prompt (as this won't change the existing behavior) and simply use the results for model/attachment events to control the value of "logoutWithUnsynedData" property we send to mixpanel.

Ok, so I have a good proof of concept branch whenever we decide to go ahead with this...

### **Extra Credit**

**Only do this if it is easy. I**

Related Stories
---------------

Card #3275 and #3276

Tasks
-----

{{table query: SELECT Number, Name, Owner, 'Task Status' WHERE Type = Task and Story = THIS CARD}}

Defects
-------

{{ table query: SELECT Number, Name, Owner, 'Status' WHERE Type = Defect and 'Related Story' = THIS CARD }}

Test Plan
---------

Enable Mixpanel on the QA site

With data:

*   Login
    *   add a Ticket or something
*   Logout

Without data:

*   Login
*   Logout

**Comment by githubbob42** [comment]
### Feature: #430 Mobile Instrumentation ([Mingle](https://liquidframeworks.mingle.thoughtworks.com/projects/alpine_mobile/cards/563))

**Comment by githubbob42** [comment]
**Pull Request #1432**
